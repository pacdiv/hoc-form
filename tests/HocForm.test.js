import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import hocForm, { Field } from '../index';
import {
  Form,
  validate,
} from './helpers';

Enzyme.configure({ adapter: new Adapter() });

const LoginForm = hocForm({
  initialValues: {
    login: 'hulk',
  },
  validate,
})(Form);

const LoginFormWithoutValidation = hocForm({})(Form);

describe('HocForm()()', () => {
  const event = {
    preventDefault: value => value,
  };
  let spy;
  const onSubmit = jest.fn(values => values);
  function submitWithNewLogin(wrapper, form, event, login, times) {
    wrapper.setState({
      values: {
        login,
      },
    }, () => {
      Promise.resolve(form.simulate('submit', event))
        .then(() => expect(spy).toHaveBeenCalledTimes(times))
        .catch(() => ({}));
    });
  }

  afterEach(() => {
    spy.mockRestore();
  });

  it('shallows HocForm with validation', () => {
    const formWrapper = shallow(
      <LoginForm
        onSubmit={onSubmit}
      />,
    );
    spy = jest.spyOn(formWrapper.instance().props, 'onSubmit');

    formWrapper.instance().setError('login', 'Please enter a login');
    formWrapper.instance().unsetError('login');
    formWrapper.instance().setValue('login', '');

    const form = formWrapper.find(Form).first().dive();
    submitWithNewLogin(formWrapper, form, event, '', 1);
    submitWithNewLogin(formWrapper, form, event, 'starman', 2);
    expect(formWrapper.render()).toMatchSnapshot();
  });

  it('shallows HocForm without sync or async validation', () => {
    const onSubmit = jest.fn(values => values)
    const formWrapper = shallow(
      <LoginFormWithoutValidation
        onSubmit={onSubmit}
      />,
    );
    const spy = jest.spyOn(formWrapper.instance().props, 'onSubmit');

    const form = formWrapper.find(Form).first().dive();
    formWrapper.instance().setValue('login', 'hulk');
    submitWithNewLogin(formWrapper, form, event, 'starman', 1);
  });
});
