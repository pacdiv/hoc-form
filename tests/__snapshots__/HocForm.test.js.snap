// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HocForm()() shallows HocForm with validation 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "novalidate": "",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "style": "display:flex;flex-direction:column;position:relative",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "style": "font-weight:300;margin-bottom:.5em",
            },
            "children": Array [
              Object {
                "data": "Login",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "label",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "name": "login",
                "placeholder": "elonmusk",
                "style": "border:1px solid lightgrey;border-color:lightgrey;border-radius:.3em;font-size:1em;font-weight:300;height:2em;margin-bottom:2em;padding:.5em 1em",
                "type": "string",
                "value": "starman",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "name": undefined,
                "placeholder": undefined,
                "style": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "name": undefined,
                "placeholder": undefined,
                "style": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "style": undefined,
            },
            "x-attribsPrefix": Object {
              "style": undefined,
            },
          },
          Object {
            "attribs": Object {
              "name": "login",
              "placeholder": "elonmusk",
              "style": "border:1px solid lightgrey;border-color:lightgrey;border-radius:.3em;font-size:1em;font-weight:300;height:2em;margin-bottom:2em;padding:.5em 1em",
              "type": "string",
              "value": "starman",
            },
            "children": Array [],
            "name": "input",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "style": "font-weight:300;margin-bottom:.5em",
              },
              "children": Array [
                Object {
                  "data": "Login",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "label",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "style": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "name": undefined,
              "placeholder": undefined,
              "style": undefined,
              "type": undefined,
              "value": undefined,
            },
            "x-attribsPrefix": Object {
              "name": undefined,
              "placeholder": undefined,
              "style": undefined,
              "type": undefined,
              "value": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "type": "submit",
          },
          "children": Array [
            Object {
              "data": "Sign in",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "type": undefined,
          },
          "x-attribsPrefix": Object {
            "type": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "type": "submit",
        },
        "children": Array [
          Object {
            "data": "Sign in",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "style": "display:flex;flex-direction:column;position:relative",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "style": "font-weight:300;margin-bottom:.5em",
              },
              "children": Array [
                Object {
                  "data": "Login",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "label",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "name": "login",
                  "placeholder": "elonmusk",
                  "style": "border:1px solid lightgrey;border-color:lightgrey;border-radius:.3em;font-size:1em;font-weight:300;height:2em;margin-bottom:2em;padding:.5em 1em",
                  "type": "string",
                  "value": "starman",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "name": undefined,
                  "placeholder": undefined,
                  "style": undefined,
                  "type": undefined,
                  "value": undefined,
                },
                "x-attribsPrefix": Object {
                  "name": undefined,
                  "placeholder": undefined,
                  "style": undefined,
                  "type": undefined,
                  "value": undefined,
                },
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "style": undefined,
              },
              "x-attribsPrefix": Object {
                "style": undefined,
              },
            },
            Object {
              "attribs": Object {
                "name": "login",
                "placeholder": "elonmusk",
                "style": "border:1px solid lightgrey;border-color:lightgrey;border-radius:.3em;font-size:1em;font-weight:300;height:2em;margin-bottom:2em;padding:.5em 1em",
                "type": "string",
                "value": "starman",
              },
              "children": Array [],
              "name": "input",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "style": "font-weight:300;margin-bottom:.5em",
                },
                "children": Array [
                  Object {
                    "data": "Login",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "label",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "style": undefined,
                },
                "x-attribsPrefix": Object {
                  "style": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "name": undefined,
                "placeholder": undefined,
                "style": undefined,
                "type": undefined,
                "value": undefined,
              },
              "x-attribsPrefix": Object {
                "name": undefined,
                "placeholder": undefined,
                "style": undefined,
                "type": undefined,
                "value": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "type": undefined,
        },
        "x-attribsPrefix": Object {
          "type": undefined,
        },
      },
    ],
    "name": "form",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "novalidate": undefined,
    },
    "x-attribsPrefix": Object {
      "novalidate": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
